aws-blog-event-driven-batch-analytics/src/main/java/com/amazonaws/bigdatablog/edba/emr/ProcessVendorTrasactions.java

```diff
@@ -0,0 +1,87 @@
+import org.apache.spark.sql.types.StructType;
+
+import com.amazonaws.auth.AWSSessionCredentials;
+import com.amazonaws.auth.InstanceProfileCredentialsProvider;
+import com.google.common.collect.ImmutableMap;
+
+import scala.collection.Seq;
+
+public class ProcessVendorTrasactions {
+
+	final static String redshiftJDBCURL="jdbc:redshift://<<YourClusterEndPoint>>:5439/test?user=<<YourUser>>&password=<<YourPassword>>";
+	public static void main(String[] args){
+		
+		    	SparkConf conf = new SparkConf().setAppName("Spark Redshift No Access-Keys");
+		    	List<StructField> schemaFields = new ArrayList<StructField>();
+		    	schemaFields.add(DataTypes.createStructField("vendor_id", DataTypes.StringType, true));
+		    	schemaFields.add(DataTypes.createStructField("trans_amount", DataTypes.StringType, true));
+		    	schemaFields.add(DataTypes.createStructField("trans_type", DataTypes.StringType, true));
+		    	schemaFields.add(DataTypes.createStructField("item_id", DataTypes.StringType, true));
+		    	schemaFields.add(DataTypes.createStructField("trans_date", DataTypes.StringType, true));
+		    	StructType schema = DataTypes.createStructType(schemaFields);
+		    	
+				JavaSparkContext sc = new JavaSparkContext(conf);
+				
+				JavaRDD<Row> salesRDD = sc.textFile(args[0]).
+						map(new Function<String,Row>(){public Row call(String saleRec){ String[] fields = saleRec.split(",");
+					      return RowFactory.create(fields[0], fields[1],fields[2],fields[3],fields[4]);}});
+				SQLContext sqlc = new SQLContext(sc);
+				DataFrame salesDF = sqlc.createDataFrame(salesRDD,schema);
+				DataFrame vendorItemSaleAmountDF = salesDF.filter(salesDF.col("trans_type").equalTo("4")).groupBy(salesDF.col("vendor_id"),salesDF.col("item_id"),salesDF.col("trans_date")).agg(ImmutableMap.of("trans_amount", "sum"));
+				DataFrame vendorItemTaxAmountDF = salesDF.filter(salesDF.col("trans_type").equalTo("5")).groupBy(salesDF.col("vendor_id"),salesDF.col("item_id"),salesDF.col("trans_date")).agg(ImmutableMap.of("trans_amount", "sum"));
+				DataFrame vendorItemDiscountAmountDF = salesDF.filter(salesDF.col("trans_type").equalTo("6")).groupBy(salesDF.col("vendor_id"),salesDF.col("item_id"),salesDF.col("trans_date")).agg(ImmutableMap.of("trans_amount", "sum"));
+				String[] joinColArray = {"vendor_id","item_id","trans_date"};
+				vendorItemSaleAmountDF.printSchema();
+				Seq<String> commonJoinColumns = scala.collection.JavaConversions.asScalaBuffer(Arrays.asList(joinColArray)).seq();
+
+				DataFrame vendorAggregatedDF = vendorItemSaleAmountDF.join(vendorItemTaxAmountDF,commonJoinColumns,"left_outer")
+										 .join(vendorItemDiscountAmountDF,commonJoinColumns,"left_outer")
+										 .toDF("vendor_id","item_id","trans_date","sale_amount","tax_amount","discount_amount");
+				vendorAggregatedDF.printSchema();
+				AWSSessionCredentials creds  = (AWSSessionCredentials) new InstanceProfileCredentialsProvider().getCredentials();
```

Comment:
For customers who run this in different ways, it's a better idea to use the DefaultCredentialsProviderChain, which will pick up InstanceProfileCredentials before anything else.